### PR TITLE
Improve error message for oversized function

### LINKED ISSUE #12608: [core] Better error message when function scope is too large
### What is the problem?
Similar to what was reported in #12603, Redis will fail if the serialized function definition is too large.

This can often be the case when adding ray decorators to python scripts.

Instead, we should:

1. Ban functions from being too large (have a check and raise a hard warning with env var override).
2. Have a debugging function for detecting scoped culprits
3. (more scifi) fallback to plasma automatically.

### PR DESCRIPTION
**Pull request by ericl**
Closes https://github.com/ray-project/ray/issues/12608

Example outputs:

ValueError: The remote function __main__.capture is too large (838862615 > FUNCTION_SIZE_ERROR_THRESHOLD=100000000 bytes). Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use ray.put() to put large objects in the Ray object store.

2021-07-15 14:19:58,063	WARNING worker.py:1188 -- Warning: The remote function __main__.capture is very large (83887895 bytes). Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use ray.put() to put large objects in the Ray object store.

### CODE REVIEW

File: python/ray/ray_constants.py [hunk 1]
```diff
@@ -69,7 +69,9 @@ def env_bool(key, default):
 
 # If a remote function or actor (or some other export) has serialized size
 # greater than this quantity, print an warning.
-PICKLE_OBJECT_WARNING_SIZE = 10**7
+FUNCTION_SIZE_WARN_THRESHOLD = 10**7
+FUNCTION_SIZE_ERROR_THRESHOLD = env_integer("FUNCTION_SIZE_ERROR_THRESHOLD",
```
Comment (ericl) [inline python/ray/ray_constants.py:73]: Empirically, Redis starts raising odd exceptions at 2-3x this threshold.

**Comment by ericl** [comment]
//python/ray/tests:test_failure                                          FAILED in 3 out of 3 in 130.1s
  | Stats over 3 runs: max = 130.1s, min = 129.1s, avg = 129.7s, dev = 0.4s
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_failure/test.log
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_failure/test_attempts/attempt_1.log
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_failure/test_attempts/attempt_2.log
 

<br class="Apple-interchange-newline">

**Comment by richardliaw** [comment]
Can you convert `838862615` to something human readable, like 838 MB?

**Comment by ericl** [comment]
Done, converted to `MiB`.

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -443,7 +443,7 @@ def debug(address):
     "--temp-dir",
     hidden=True,
     default=None,
-    help="manually specify the root temporary dir of the Ray process")
+    help="Deprecated: use the TMPDIR env var instead.")
```
Comment (richardliaw) [inline python/ray/scripts/scripts.py:446]: revert? up to u

**Review by richardliaw — APPROVED** [review]

File: python/ray/worker.py [hunk 3]
```diff
@@ -683,9 +683,7 @@ def init(
         _memory: Amount of reservable memory resource to create.
         _redis_password (str): Prevents external clients without the password
             from connecting to Redis if provided.
-        _temp_dir (str): If provided, specifies the root temporary
-            directory for the Ray process. Defaults to an OS-specific
-            conventional location, e.g., "/tmp/ray".
+        _temp_dir (str): Deprecated: please use the TMPDIR env var instead.
```
Comment (richardliaw) [inline python/ray/worker.py:686]: same here

**Pull request merged**